File: clients/v2boxkit_client/lib/src/services/xray_config_builder.dart
import 'dart:convert';
import 'dart:io';

import '../model/models.dart';

class XrayConfigBuilder {
  const XrayConfigBuilder();

  static const _privateNetworks = <String>[
    '0.0.0.0/8',
    '10.0.0.0/8',
    '100.64.0.0/10',
    '127.0.0.0/8',
    '169.254.0.0/16',
    '172.16.0.0/12',
    '192.168.0.0/16',
    '224.0.0.0/4',
    '::1/128',
    'fc00::/7',
    'fe80::/10',
    'ff00::/8',
  ];

  Map<String, Object?> build(
    Map<String, Object?> base,
    RoutingSettings settings,
  ) {
    final configuration = (jsonDecode(jsonEncode(base)) as Map)
        .cast<String, Object?>();
    final rawOutbounds = configuration['outbounds'];
    if (rawOutbounds is! List || rawOutbounds.isEmpty) {
      throw const FormatException('节点没有生成可用的 Xray 出站配置');
    }

    final outbounds = rawOutbounds
        .map((item) => (item as Map).cast<String, Object?>())
        .where((item) => !const {'direct', 'block'}.contains(item['tag']))
        .toList();
    if (outbounds.isEmpty) {
      throw const FormatException('节点没有生成可用的 Xray 出站配置');
    }
    for (final outbound in outbounds) {
      // libXray stores the share-link display name in sendThrough, while
      // Xray-core interprets that field as a source address at runtime.
      outbound.remove('sendThrough');
    }

    outbounds.first['tag'] = 'proxy';
    outbounds.addAll([
      <String, Object?>{'tag': 'direct', 'protocol': 'freedom'},
      <String, Object?>{'tag': 'block', 'protocol': 'blackhole'},
    ]);

    configuration
      ..['log'] = <String, Object?>{'loglevel': 'warning'}
      ..['outbounds'] = outbounds
      ..['inbounds'] = _inbounds(settings)
      ..['dns'] = <String, Object?>{
        'servers': [settings.dnsServer],
        'queryStrategy': 'UseIP',
      }
      ..['routing'] = <String, Object?>{
        'domainStrategy': 'IPIfNonMatch',
        'rules': _routingRules(settings),
      };

    return configuration;
  }

  List<Map<String, Object?>> _inbounds(RoutingSettings settings) {
    final tunSettings = <String, Object?>{
      'name': Platform.isWindows ? 'V2BoxKit' : 'v2boxkit',
      'mtu': settings.mtu,
    };
    if (Platform.isWindows) {
      tunSettings.addAll({
        'desc': 'V2BoxKit',
        'gateway': ['172.19.0.1/30', 'fd00:172:19::1/64'],
        'dns': [_plainDnsAddress(settings.dnsServer)],
        'autoSystemRoutingTable': ['0.0.0.0/0', '::/0'],
        'autoOutboundsInterface': 'auto',
      });
    }

    final inbounds = <Map<String, Object?>>[
      {
        'tag': 'tun-in',
        'protocol': 'tun',
        'settings': tunSettings,
        'sniffing': {
          'enabled': true,
          'destOverride': ['http', 'tls', 'quic'],
        },
      },
    ];

    final sharing = settings.localSharing;
    if (sharing.isEnabled) {
      if (!sharing.isValid) {
        throw const FormatException('本地代理分享必须设置用户名和密码');
      }
      final accounts = [
        {'user': sharing.username, 'pass': sharing.password},
      ];
      inbounds.addAll([
        {
          'tag': 'local-socks-in',
          'listen': sharing.listenAddress,
          'port': sharing.socksPort,
          'protocol': 'socks',
          'settings': {'auth': 'password', 'accounts': accounts, 'udp': true},
          'sniffing': {
            'enabled': true,
            'destOverride': ['http', 'tls', 'quic'],
          },
        },
        {
          'tag': 'local-http-in',
          'listen': sharing.listenAddress,
          'port': sharing.httpPort,
          'protocol': 'http',
          'settings': {'accounts': accounts},
          'sniffing': {
            'enabled': true,
            'destOverride': ['http', 'tls'],
          },
        },
      ]);
    }
    return inbounds;
  }

  List<Map<String, Object?>> _routingRules(RoutingSettings settings) {
    switch (settings.mode) {
      case RoutingMode.global:
        return const [];
      case RoutingMode.direct:
        return const [
          {'type': 'field', 'network': 'tcp,udp', 'outboundTag': 'direct'},
        ];
      case RoutingMode.rule:
        final rules = <Map<String, Object?>>[];
        for (final rule in settings.customRules) {
          final converted = _customRule(rule);
          if (converted != null) rules.add(converted);
        }
        if (settings.bypassPrivateNetworks) {
          rules.add({
            'type': 'field',
            'ip': _privateNetworks,
            'outboundTag': 'direct',
          });
        }
        final domains = settings.directDomains
            .map((domain) => domain.trim().toLowerCase())
            .where((domain) => domain.isNotEmpty)
            .map(
              (domain) => domain.startsWith(RegExp(r'(domain|full|regexp):'))
                  ? domain
                  : 'domain:$domain',
            )
            .toList();
        if (domains.isNotEmpty) {
          rules.add({
            'type': 'field',
            'domain': domains,
            'outboundTag': 'direct',
          });
        }
        return rules;
    }
  }

  Map<String, Object?>? _customRule(RoutingRule rule) {
    if (!rule.isEnabled || rule.normalizedValues.isEmpty) return null;
    final outboundTag = switch (rule.action) {
      RoutingRuleAction.proxy => 'proxy',
      RoutingRuleAction.direct => 'direct',
      RoutingRuleAction.block => 'block',
    };
    final result = <String, Object?>{
      'type': 'field',
      'outboundTag': outboundTag,
    };
    switch (rule.kind) {
      case RoutingRuleKind.domain:
        result['domain'] = rule.normalizedValues.map((value) {
          final normalized = value.toLowerCase();
          return normalized.startsWith(RegExp(r'(domain|full|regexp):'))
              ? normalized
              : 'domain:$normalized';
        }).toList();
      case RoutingRuleKind.ip:
        result['ip'] = rule.normalizedValues;
    }
    return result;
  }

  String _plainDnsAddress(String value) {
    final normalized = value.trim();
    if (InternetAddress.tryParse(normalized) != null) return normalized;
    final uri = Uri.tryParse(normalized);
    if (uri != null && InternetAddress.tryParse(uri.host) != null) {
      return uri.host;
    }
    return '1.1.1.1';
  }
}

File: Sources/Shared/XrayConfigBuilder.swift
import Foundation

enum XrayConfigBuilder {
  enum BuilderError: LocalizedError {
    case invalidBaseConfiguration
    case missingOutbound
    case serializationFailed

    var errorDescription: String? {
      switch self {
      case .invalidBaseConfiguration: "libXray 返回了无法识别的配置"
      case .missingOutbound: "配置中没有可用的出站节点"
      case .serializationFailed: "Xray 配置序列化失败"
      }
    }
  }

  static func build(
    baseConfiguration: Any,
    tunnelFileDescriptor: Int32,
    routing: RoutingSettings
  ) throws -> String {
    guard var configuration = baseConfiguration as? [String: Any] else {
      throw BuilderError.invalidBaseConfiguration
    }
    guard var outbounds = configuration["outbounds"] as? [[String: Any]],
      !outbounds.isEmpty
    else {
      throw BuilderError.missingOutbound
    }

    outbounds[0]["tag"] = "proxy"
    let managedTags = Set(["direct", "block"])
    outbounds.removeAll { outbound in
      guard let tag = outbound["tag"] as? String else { return false }
      return managedTags.contains(tag)
    }
    outbounds.append(["tag": "direct", "protocol": "freedom"])
    outbounds.append(["tag": "block", "protocol": "blackhole"])

    configuration["outbounds"] = outbounds
    var inbounds: [[String: Any]] = [
      [
        "tag": "tun-in",
        "protocol": "tun",
        "settings": [
          "name": "utun",
          "mtu": routing.mtu,
        ],
        "sniffing": [
          "enabled": true,
          "destOverride": ["http", "tls", "quic"],
        ],
      ]
    ]
    if routing.localSharing.isEnabled, routing.localSharing.isValid {
      let sharing = routing.localSharing
      let accounts = [["user": sharing.username, "pass": sharing.password]]
      inbounds.append([
        "tag": "local-socks-in",
        "listen": sharing.listenAddress,
        "port": sharing.socksPort,
        "protocol": "socks",
        "settings": [
          "auth": "password",
          "accounts": accounts,
          "udp": true,
        ],
        "sniffing": [
          "enabled": true,
          "destOverride": ["http", "tls", "quic"],
        ],
      ])
      inbounds.append([
        "tag": "local-http-in",
        "listen": sharing.listenAddress,
        "port": sharing.httpPort,
        "protocol": "http",
        "settings": ["accounts": accounts],
        "sniffing": [
          "enabled": true,
          "destOverride": ["http", "tls"],
        ],
      ])
    }
    configuration["inbounds"] = inbounds
    configuration["env"] = [
      "xray.tun.fd": String(tunnelFileDescriptor)
    ]
    configuration["log"] = ["loglevel": "warning"]
    configuration["routing"] = [
      "domainStrategy": "IPIfNonMatch",
      "rules": routingRules(for: routing),
    ]

    guard JSONSerialization.isValidJSONObject(configuration),
      let data = try? JSONSerialization.data(withJSONObject: configuration)
    else {
      throw BuilderError.serializationFailed
    }
    return String(decoding: data, as: UTF8.self)
  }

  private static func routingRules(for routing: RoutingSettings) -> [[String: Any]] {
    switch routing.mode {
    case .global:
      return []

    case .direct:
      return [
        [
          "type": "field",
          "network": "tcp,udp",
          "outboundTag": "direct",
        ]
      ]

    case .rule:
      var rules: [[String: Any]] = []
      rules.append(contentsOf: customRoutingRules(routing.customRules))
      if routing.bypassPrivateNetworks {
        rules.append([
          "type": "field",
          "ip": privateNetworkCIDRs,
          "outboundTag": "direct",
        ])
      }
      let domains = routing.normalizedDirectDomains.map { "domain:\($0)" }
      if !domains.isEmpty {
        rules.append([
          "type": "field",
          "domain": domains,
          "outboundTag": "direct",
        ])
      }
      return rules
    }
  }

  private static func customRoutingRules(_ rules: [RoutingRule]) -> [[String: Any]] {
    rules.compactMap { rule in
      guard rule.isEnabled, !rule.normalizedValues.isEmpty else { return nil }

      let outboundTag: String
      switch rule.action {
      case .proxy: outboundTag = "proxy"
      case .direct: outboundTag = "direct"
      case .block: outboundTag = "block"
      }

      var result: [String: Any] = [
        "type": "field",
        "outboundTag": outboundTag,
      ]
      switch rule.kind {
      case .domain:
        result["domain"] = rule.normalizedValues.map { value in
          if value.hasPrefix("domain:") || value.hasPrefix("full:") || value.hasPrefix("regexp:") {
            return value
          }
          return "domain:\(value.lowercased())"
        }
      case .ip:
        result["ip"] = rule.normalizedValues
      }
      return result
    }
  }

  private static let privateNetworkCIDRs = [
    "0.0.0.0/8",
    "10.0.0.0/8",
    "100.64.0.0/10",
    "127.0.0.0/8",
    "169.254.0.0/16",
    "172.16.0.0/12",
    "192.168.0.0/16",
    "224.0.0.0/4",
    "::1/128",
    "fc00::/7",
    "fe80::/10",
    "ff00::/8",
  ]
}
